File: games.R
# open up a graphics window
windows()

# ----------------------------------------------------------------------------------
# ---- Packages --------------------------------------------------------------------
# ----------------------------------------------------------------------------------

require(ggplot2)
require(scales)
require(stringr)
require(data.table)
require(foreach)
require(doSNOW)

# ----------------------------------------------------------------------------------
# ---- Functions -------------------------------------------------------------------
# ----------------------------------------------------------------------------------

# build a function for rolling dice
roll_dice = function(rolls = 1000, dice = 2, sides = 6, seed = NULL)
{
  require(data.table)
  
  if(!is.null(seed)) set.seed(seed)
  results = data.table(t(sapply(1:rolls, function(i) sample(x = 1:sides, size = dice, replace = TRUE))))
  setnames(results, paste0("Dice_", 1:ncol(results)))
  
  return(results)
}

# build a function for drawing cards
draw_cards = function(decks = 1, draws = 18, seed = NULL)
{
  require(data.table)
  require(stringr)
  
  cards = unlist(lapply(c("Hearts", "Diamonds", "Spades", "Clubs"), function(i) paste(c(2:10, "Jack", "King", "Queen", "Ace"), i, sep = "_")))
  
  if(!is.null(seed)) set.seed(seed)
  drawing = sample(x = rep(cards, decks), size = draws, replace = FALSE)
  
  drawing = data.table(str_split_fixed(string = drawing, pattern = "_", n = 2))
  setnames(drawing, c("Card", "Suite"))
  
  return(drawing)
}

# build a function for spinning roulette
spin_roulette = function(spins = 1000, seed = NULL)
{
  require(data.table)
  
  values = data.table(Number = c("00", as.character(0:36)),
                      First_12 = c(rep("No", 2), rep("Yes", 12), rep("No", 24)),
                      Second_12 = c(rep("No", 14), rep("Yes", 12), rep("No", 12)),
                      Third_12 = c(rep("No", 26), rep("Yes", 12)),
                      First_18 = c(rep("No", 2), rep("Yes", 18), rep("No", 18)),
                      Second_18 = c(rep("No", 20), rep("Yes", 18)),
                      Even = c(rep("No", 2), rep(c("No", "Yes"), 18)),
                      Odd = c(rep("No", 2), rep(c("Yes", "No"), 18)),
                      Green = c(rep("Yes", 2), rep("No", 36)),
                      Red = c(rep("No", 2), rep(c(rep(c("Yes", "No"), 5)[-10], c("No", "No", "Yes", "No", "Yes", "No", "Yes", "No", "Yes")), 2)))
  
  values[, Black := ifelse(Red == "Yes", "No", ifelse(Green == "Yes", "No", "Yes"))]
  values[, Min_2to1 := c(rep("No", 2), rep(c("Yes", "No", "No"), 12))]
  values[, Mid_2to1 := c(rep("No", 2), rep(c("No", "Yes", "No"), 12))]
  values[, Max_2to1 := c(rep("No", 2), rep(c("No", "No", "Yes"), 12))]
  
  if(!is.null(seed)) set.seed(seed)
  turns = sample(x = 1:nrow(values), size = spins, replace = TRUE)
  
  return(values[turns])
}

# check out the output of each function
roll_dice()
draw_cards()
spin_roulette()

# ----------------------------------------------------------------------------------
# ---- Rolling Dice ----------------------------------------------------------------
# ----------------------------------------------------------------------------------

# roll a pair of dice
dice = roll_dice(rolls = 3e5, dice = 2, sides = 6, seed = 21)

# compute the total of each roll, and the combination of each roll
dice[, Total := rowSums(dice)]
dice[, Combo := paste0(ifelse(Dice_1 < Dice_2, Dice_1, Dice_2), ",", ifelse(Dice_1 > Dice_2, Dice_1, Dice_2))]

# plot the distribution of the dice roll total
plot_total = ggplot(data = dice, aes(x = Total)) + 
  geom_histogram(fill = "cornflowerblue", color = "white", binwidth = 1) + 
  scale_x_continuous(breaks = sort(unique(dice$Total))) + 
  scale_y_continuous(label = comma) + 
  ggtitle("Rolling Two Dice") + 
  labs(x = "Total", y = "Frequency") + 
  theme_bw(base_size = 25) +
  theme(legend.position = "none", 
        legend.key.size = unit(.25, "in"),
        plot.title = element_text(hjust = 0.5),
        # axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  guides(color = guide_legend(override.aes = list(size = 10, linetype = 1, alpha = 1), nrow = 1, byrow = TRUE))

plot_total

# plot the distribution of the dice roll combination
plot_combo = ggplot(data = dice, aes(x = as.factor(Combo))) + 
  geom_bar(fill = "cornflowerblue", color = "white") + 
  scale_y_continuous(label = comma) + 
  ggtitle("Rolling Two Dice") + 
  labs(x = "Combination", y = "Frequency") + 
  theme_bw(base_size = 25) +
  theme(legend.position = "none", 
        legend.key.size = unit(.25, "in"),
        plot.title = element_text(hjust = 0.5),
        # axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank()) +
  guides(color = guide_legend(override.aes = list(size = 10, linetype = 1, alpha = 1), nrow = 1, byrow = TRUE))

plot_combo

# ----------------------------------------------------------------------------------
# ---- Blackjack -------------------------------------------------------------------
# ----------------------------------------------------------------------------------

# blackjack rules: https://www.wikihow.com/Play-Blackjack

# how many players are there?
players = 4

# what are the values for players standing?
player_stands = 12:17

# what is the value for the dealer standing?
dealer_stands = 17

# how many hands will be played?
hands = 100

# how many decks will be used?
decks = 2

# create standing strategies
stand_strategies = data.table(expand.grid(append(lapply(1:players, function(i) player_stands), list(dealer_stands))))

# update the column names for stand_strategies
setnames(stand_strategies, c(paste0("Player_", 1:players), "Dealer"))

# create a table of values for each card
card_values = data.table(Card = c(2:10, "Jack", "Queen", "King", "Ace"), 
                         Value = c(2:10, 10, 10, 10, 11))

# draw cards for each hand
set.seed(42)
draw_hands = lapply(1:hands, function(i) draw_cards(decks = decks, draws = 52 * decks))

# setup a cluster for parallel processing
# workers = getDTthreads() - 1
# cl = makeCluster(workers, type = "SOCK", outfile = "")
# registerDoSNOW(cl)

# play through each hand
# blackjack_success = foreach(i = 1:length(draw_hands)) %dopar%
blackjack_success = lapply(1:length(draw_hands), function(i)
{
  require(data.table)
  
  # get hand i and keep track of its drawing order
  table_hands = data.table(draw_hands[[i]])
  table_hands[, index := 1:nrow(table_hands)]
  
  # join card_values onto table_hands
  setkey(table_hands, Card)
  setkey(card_values, Card)
  table_hands = card_values[table_hands]
  table_hands = table_hands[order(index)]
  table_hands[, index := NULL]
  
  # assign cards to players
  table_hands[, Player := c(rep(c(paste0("Player_", 1:players), "Dealer"), 2), rep("Unknown", nrow(table_hands) - ((players + 1) * 2)))]
  
  # create the order of Players
  player_order = c(paste0("Player_", 1:players), "Dealer")
  
  # split up table_hands by player_order
  player_hands = lapply(player_order, function(p) data.table(table_hands[Player == p]))
  
  # get the undealt cards
  undealt_cards = data.table(table_hands[Player == "Unknown"])
  
  # play through each strategy on hand i
  strategy_success = unlist(lapply(1:nrow(stand_strategies), function(k)
  {
    # set up an empty vector to store hand values for strategy k
    hand_values = c()
    
    # play through strategy k for each player
    for(j in 1:length(player_hands))
    {
      # get player j's hand
      player_hand = data.table(player_hands[[j]])
      
      # get the value of player j's hand
      hand_value = sum(player_hand$Value)
      
      # get the value of player j's k-th stand_strategy
      stand_value = unname(unlist(stand_strategies[k, j, with = FALSE]))
      
      # draw another card while hand_value is less than stand_value
      while(hand_value < stand_value)
      {
        # add a card to player_hand
        player_hand = rbind(player_hand, undealt_cards[1])
        
        # remove a card from undealt_cards
        undealt_cards = undealt_cards[-1]
        
        # update hand_value
        hand_value = sum(player_hand$Value)
        
        # if the hand_value broke 21 and there's an Ace in the hand, update the Ace value to 1
        if(hand_value > 21 & "Ace" %in% player_hand$Card)
        {
          # update the value of Ace
          player_hand[Card == "Ace", Value := 1]
          
          # update hand_value
          hand_value = sum(player_hand$Value)
        }
      }
      
      # add hand_value to hand_values
      hand_values = append(hand_values, hand_value)
    }
    
    # determine the table win percentage for strategy k on hand i
    win_percent = sum(unlist(lapply(1:players, function(m)
      ifelse(hand_values[players + 1] > 21 & hand_values[m] <= 21, 1, 
             ifelse(hand_values[m] > hand_values[players + 1] & hand_values[m] <= 21, 1, 0))))) / players
    
    return(win_percent)
  }))
  
  # make strategy_success into a table
  strategy_success = cbind(data.table(Hand = i, Strategy = 1:length(strategy_success), Success = strategy_success), stand_strategies)
  
  # print out completion of hand i
  cat("----", as.character(nrow(stand_strategies)), "strategies on hand", i, "of", as.character(length(draw_hands)), "evaluated at", as.character(Sys.time()), "----\n")
  
  return(strategy_success)
})

# end the cluster
# stopCluster(cl)

# combine the list of tables into one table
blackjack_success = rbindlist(blackjack_success)

# average the success of each strategy
strategy_score = data.table(blackjack_success[, .(Success = mean(Success)), by = .(Strategy)])

# rank the success of each strategy
strategy_score = strategy_score[order(Success, decreasing = TRUE)]

# get the average stand value
strategy_score[, Stand_Avg := rowSums(stand_strategies[strategy_score$Strategy, 1:players, with = FALSE]) / players]

# add the player stand values to strategy_score
strategy_score = cbind(strategy_score, stand_strategies[strategy_score$Strategy, 1:players, with = FALSE])

# see if there is a linear relationship between stand strategy and success
cor(strategy_score$Success, strategy_score$Stand_Avg)
cor(strategy_score$Success, strategy_score$Player_1)
cor(strategy_score$Success, strategy_score$Player_2)
cor(strategy_score$Success, strategy_score$Player_3)
cor(strategy_score$Success, strategy_score$Player_4)
strategy_score

# ----------------------------------------------------------------------------------
# ---- Craps -----------------------------------------------------------------------
# ----------------------------------------------------------------------------------

# craps rules: https://www.wikihow.com/Play-Craps




# ----------------------------------------------------------------------------------
# ---- Roulette --------------------------------------------------------------------
# ----------------------------------------------------------------------------------

# roulette rules: https://www.wikihow.com/Play-Roulette




# ----------------------------------------------------------------------------------
# ---- 5-Card Poker ----------------------------------------------------------------
# ----------------------------------------------------------------------------------

# 5-card poker rules: https://www.wikihow.com/Play-Five-Card-Draw

File: games.py
# set the path of where the input files are
mywd = "C:\\Users\\Nick Morris\\Desktop"

# -----------------------------------------------------------------------------------
# ---- Packages ---------------------------------------------------------------------
# -----------------------------------------------------------------------------------

# data handling
import os
import pandas as pd
import numpy as np
import gc
import time

# graphics
from plotnine import *

# ----------------------------------------------------------------------------------
# ---- Functions -------------------------------------------------------------------
# ----------------------------------------------------------------------------------

# build a function for rolling dice
def roll_dice(rolls = 100, dice = 2, sides = 6, seed = None):
    
    # set the seed
    np.random.seed(seed)
    
    # roll the dice
    dice_rolls = pd.DataFrame(np.random.choice(a = [i + 1 for i in range(sides)],
                                               size = (rolls, dice),
                                               replace = True),
                              columns = ["Dice_" + str(i + 1) for i in range(dice)])
    return dice_rolls

# build a function for drawing cards
def draw_cards(decks = 1, draws = 18, seed = None):
    
    # set up all the suites
    suites = ["Hearts", "Diamonds", "Spades", "Clubs"]
    
    # set up all the face values
    faces = [i + 2 for i in range(9)]
    faces.extend(["Jack", "King", "Queen", "Ace"])
    
    # assign each face value to each suite
    cards = [str(j) + "_" + i for j in faces for i in suites]
    
    # set the seed
    np.random.seed(seed)
    
    # create decks of cards and shuffle them
    cards = np.random.permutation(np.repeat(cards, decks))
    
    # draw the cards
    drawing = pd.Series([cards[i] for i in np.random.choice(a = range(len(cards)),
                                                            size = (min(draws, len(cards))),
                                                            replace = False)])
    
    # split up the cards by their face value and suite
    drawing = drawing.str.split("_", expand = True)
    
    # update the column names in drawing
    drawing.columns = ["Face", "Suite"]
    
    return drawing

# build a function for spinning roulette
def spin_roulette(spins = 100, seed = None):
    
    # set up all the numbers
    Number = np.append(["00"], [str(i) for i in range(37)])
    
    # determine which numbers are the first 12
    First_12 = np.concatenate((np.repeat("No", 2), np.repeat("Yes", 12), np.repeat("No", 24)))
    
    # determine which numbers are the second 12
    Second_12 = np.concatenate((np.repeat("No", 14), np.repeat("Yes", 12), np.repeat("No", 12)))
    
    # determine which numbers are the third 12
    Third_12 = np.concatenate((np.repeat("No", 26), np.repeat("Yes", 12)))
    
    # determine which numbers are the first 18
    First_18 = np.concatenate((np.repeat("No", 2), np.repeat("Yes", 18), np.repeat("No", 18)))
    
    # determine which numbers are the second 18
    Second_18 = np.concatenate((np.repeat("No", 20), np.repeat("Yes", 18)))
    
    # determine which numbers are even
    Even = np.concatenate((np.repeat("No", 2), np.tile(["No", "Yes"], 18)))
    
    # determine which numbers are odd
    Odd = np.concatenate((np.repeat("No", 2), np.tile(["Yes", "No"], 18)))
    
    # determine which numbers are green
    Green = np.concatenate((np.repeat("Yes", 2), np.repeat("No", 36)))
    
    # determine which numbers are red
    Red = np.concatenate((np.repeat("No", 2), np.tile(np.append(np.tile(["Yes", "No"], 5)[:-1], ["No", "No", "Yes", "No", "Yes", "No", "Yes", "No", "Yes"]), 2)))
    
    # determine which numbers are black
    Black = ["Yes" if i == "No" else "No" for i in Red]
    Black[:2] = ["No", "No"]
    Black = np.array(Black)
    
    # determine which numbers are low 2 to 1
    Low_2to1 = np.concatenate((np.repeat("No", 2), np.tile(["Yes", "No", "Yes"], 12)))
    
    # determine which numbers are middle 2 to 1
    Middle_2to1 = np.concatenate((np.repeat("No", 2), np.tile(["No", "Yes", "No"], 12)))
    
    # determine which numbers are high 2 to 1
    High_2to1 = np.concatenate((np.repeat("No", 2), np.tile(["No", "No", "Yes"], 12)))
    
    # create the roulette values
    roulette = pd.DataFrame({"Number": Number,
                             "First_12": First_12,
                             "Second_12": Second_12,
                             "Third_12": Third_12,
                             "First_18": First_18,
                             "Second_18": Second_18,
                             "Even": Even,
                             "Odd": Odd,
                             "Green": Green,
                             "Red": Red,
                             "Black": Black,
                             "Low_2to1": Low_2to1,
                             "Middle_2to1": Middle_2to1,
                             "High_2to1": High_2to1})
    
    # spin values from roulette 
    values = roulette.sample(n = spins, axis = 0, replace = True, random_state = seed).reset_index(drop = True)
    
    return values

# check out the output of each function
roll_dice()
draw_cards()
spin_roulette()

# ----------------------------------------------------------------------------------
# ---- Rolling Dice ----------------------------------------------------------------
# ----------------------------------------------------------------------------------

# set the work directory
os.chdir(mywd)

# roll a pair of dice
dice = roll_dice(rolls = int(3e5), dice = 2, sides = 6, seed = 21)

# compute the total of each roll
dice["Total"] = dice.sum(axis = 1)

# compute the combination of each roll
dice["Combo"] = dice[["Dice_1", "Dice_2"]].values.tolist()

# sort each roll combination
dice["Combo"] = dice["Combo"].apply(sorted)

# make Combo categorical
dice["Combo"] = pd.Categorical(values = dice["Combo"].astype("str"))

# plot the distribution of the dice roll total
total_plot = (ggplot(dice, aes(x = "Total")) +
  geom_histogram(fill = "cornflowerblue", color = "white", binwidth = 1) + 
  scale_x_continuous(breaks = tuple(set(dice["Total"]))) + 
  scale_y_continuous(labels = lambda l: [format(int(np.round(v, 0)), ",") for v in l]) +
  ggtitle("Rolling Two Dice\n") +
  labs(x = "Total", y = "Frequency") +
  theme_bw(25) +
  theme(plot_title = element_text(hjust = 0.5, vjust = 1),
        figure_size = (14, 10),
        aspect_ratio = 4/5,
        panel_grid_major = element_blank(),
        panel_grid_minor = element_blank()))

total_plot

# plot the distribution of the dice roll combination
combo_plot = (ggplot(dice, aes(x = "Combo")) +
  geom_bar(fill = "cornflowerblue", color = "white") + 
  scale_y_continuous(labels = lambda l: [format(int(np.round(v, 0)), ",") for v in l]) +
  ggtitle("Rolling Two Dice\n") +
  labs(x = "Combination", y = "Frequency") +
  theme_bw(25) +
  theme(plot_title = element_text(hjust = 0.5, vjust = 1),
        figure_size = (14, 10),
        aspect_ratio = 4/5,
        axis_text_x = element_text(angle = 45, hjust = 0.5, vjust = 1),
        panel_grid_major = element_blank(),
        panel_grid_minor = element_blank()))

combo_plot

# ----------------------------------------------------------------------------------
# ---- Blackjack -------------------------------------------------------------------
# ----------------------------------------------------------------------------------

# blackjack rules: https://www.wikihow.com/Play-Blackjack

# how many players are there?
players = 5

# what are the values for players standing?
player_stands = [i for i in range(12, 17)]

# what is the value for the dealer standing?
dealer_stands = [17]

# how many hands will be played?
hands = 1000

# how many decks will be used?
decks = 7

# set up a grid for standing strategies
stand_strategies = pd.DataFrame(np.array(np.meshgrid(player_stands,
                                                     player_stands,
                                                     player_stands,
                                                     player_stands,
                                                     player_stands,
                                                     dealer_stands)).reshape(players + 1, int(len(player_stands)**players * len(dealer_stands))).T,
                                columns = np.concatenate((["Player_" + str(i + 1) for i in range(players)], ["Dealer"])))

# add a Strategy column to stand_strategies
stand_strategies["Strategy"] = stand_strategies.index.values

# set up all the face values
faces = [str(i + 2) for i in range(9)]
faces.extend(["Jack", "King", "Queen", "Ace"])

# set up all the point values for each face
points = [i + 2 for i in range(9)]
points.extend([10, 10, 10, 11])

# create a table of face values
face_values = pd.DataFrame({"Face": faces, "Value": points})

# draw cards for each hand
draw_hands = []
np.random.seed(42)
seeds = np.random.choice(a = [i + 1 for i in range(int(hands * 100))],
                         size = (hands),
                         replace = False)
for i in range(hands):
    draw_hands.append(draw_cards(decks = decks, draws = (players + 1) * (2 + 4), seed = seeds[i]))

# create the order of Players
player_order = np.concatenate((["Player_" + str(o + 1) for o in range(players)], ["Dealer"]))

# set up a table to hold the strategy success
blackjack_success = pd.DataFrame(columns = np.append(["Hand", "Strategy"], [w + "_won" for w in player_order[:-1]]))

# play through each hand using stand_strategies
for i in range(214, hands):
    
    # play through each strategy on hand i
    for k in stand_strategies.index.values:
        
        # get hand i and keep track of its drawing order
        table_hands = draw_hands[i]
        table_hands = table_hands.reset_index()
        
        # join face_values onto table_hands
        table_hands = pd.merge(face_values, table_hands, left_on = "Face", right_on = "Face", how = "right")
        
        # sort table_hands by index
        table_hands = table_hands.sort_values(by = "index", ascending = True).reset_index(drop = True)
        
        # remove the index column
        table_hands = table_hands.drop("index", axis = 1)
        
        # deal cards to players
        table_hands["Player"] = np.concatenate((np.tile(np.concatenate((["Player_" + str(d + 1) for d in range(players)], ["Dealer"])), 2),
                                                np.tile(["Unknown"], len(table_hands.index) - ((players + 1) * 2))))
        
        # split up table_hands by player_order
        player_hands = []
        for j in player_order:
            player_hands.append(table_hands.loc[table_hands["Player"] == j].reset_index(drop = True))
        
        # get the undealt cards
        undealt_cards = table_hands.loc[table_hands["Player"] == "Unknown"].reset_index(drop = True)
        
        # set up a list to hold the final value of each players hand
        hand_values = []
        
        # play through strategy k for each player
        for p in range(len(player_order)):
            
            # get player p's hand
            player_hand = player_hands[p]
            
            # get the value of player p's hand
            hand_value = sum(player_hand["Value"])
            
            # get the value of player p's k-th stand_strategy
            stand_value = stand_strategies[player_order[p]][k]
            
            # draw another card while hand_value is less than stand_value
            while hand_value < stand_value:
                
                # add a card to player_hand
                player_hand = pd.concat([player_hand, undealt_cards.head(1)], axis = "rows").reset_index(drop = True)
                
                # remove this card from undealt_cards
                undealt_cards = undealt_cards.iloc[1:].reset_index(drop = True)
                
                # get the value of player p's hand
                hand_value = sum(player_hand["Value"])
                
                # if hand_value > 21 and there's an Ace in player_hand, update the Ace value to 1
                if (hand_value > 21 and "Ace" in player_hand["Face"]):
                    
                    # determine where the Ace is
                    ace_position = np.where(np.array(player_hand["Face"]) == "Ace")[0].astype("int")
                    
                    # update the Ace value(s)
                    for a in ace_position:
                        player_hand.at[ace_position[a], "Value"] = 1
                    
                    # get the value of player p's hand
                    hand_value = sum(player_hand["Value"])
                    
            # add hand_value to hand_values
            hand_values = np.append(hand_values, hand_value)
        
        # set up a list for holding strategy success
        strategy_success = [i, k]
        
        # determine the table results for strategy k on hand i
        for r in range(players):
            
            # if the dealer broke 21 and player r didn't, that's a win
            if (hand_values[players] > 21 and hand_values[r] <= 21):
                strategy_success = np.append(strategy_success, [1])
            
            # if player r beat the dealer without breaking 21, that's a win
            elif (hand_values[r] > hand_values[players] and hand_values[r] <= 21):
                strategy_success = np.append(strategy_success, [1])
            
            # otherwise player r lost
            else:
                strategy_success = np.append(strategy_success, [0])
        
        # convert strategy_success into a table
        strategy_success_table = pd.DataFrame(columns = blackjack_success.columns)
        strategy_success_table.loc[0] = strategy_success
        
        # add strategy_success_table to blackjack_success
        blackjack_success = pd.concat([blackjack_success, strategy_success_table], axis = "rows").reset_index(drop = True)
        
        # clean out the garbage in RAM
        gc.collect()
        
        # report progress
        print("---- Blackjack Strategy " + str(k + 1) + " of " + str(len(stand_strategies.index.values)) + " on Hand " + str(i + 1) + " of " + str(hands) + " completed on " + time.ctime() + " ----")

# export the results
# blackjack_success.to_csv("Blackjack Simulation - Part 3.csv", index = False)

# import the results
# blackjack_success1 = pd.read_csv("Blackjack Simulation - Part 1.csv")
# blackjack_success2 = pd.read_csv("Blackjack Simulation - Part 2.csv")
# blackjack_success = pd.concat([blackjack_success1, blackjack_success2], axis = "index").reset_index(drop = True)

# compute the starting value of each players hand
start_values = pd.DataFrame(columns = np.append(["Hand"], [h + "_hand" for h in player_order[:-1]]))
for i in range(hands):
    
    # get hand i and keep track of its drawing order
    hand_i = draw_hands[i]
    hand_i = hand_i.reset_index()
    
    # join face_values onto hand_i
    hand_i = pd.merge(face_values, hand_i, left_on = "Face", right_on = "Face", how = "right")
    
    # sort hand_i by index
    hand_i = hand_i.sort_values(by = "index", ascending = True).reset_index(drop = True)
    
    # remove the index column
    hand_i = hand_i.drop("index", axis = 1)
    
    # deal cards to players
    hand_i["Player"] = np.concatenate((np.tile(np.concatenate((["Player_" + str(d + 1) for d in range(players)], ["Dealer"])), 2),
                                            np.tile(["Unknown"], len(hand_i.index) - ((players + 1) * 2))))
    
    # compute the starting value of hand_i for each player
    hand_i_values = [i]
    for j in player_order[:-1]:
        hand_i_values = np.append(hand_i_values, sum(hand_i.loc[hand_i["Player"] == j]["Value"]))
    
    # store hand_i_values
    start_values.loc[i] = hand_i_values

# compute the average table strategy
stand_strategies["Table"] = np.round(stand_strategies[player_order[:-1]].apply(np.mean, axis = 1), 1)

# compute the table win percentage
blackjack_success["Table_won"] = np.divide(blackjack_success[[w + "_won" for w in player_order[:-1]]].apply(np.sum, axis = 1), len(player_order[:-1]))

# create an index column in blackjack_success
blackjack_success = blackjack_success.reset_index()

# join stand_strategies onto blackjack_success
blackjack_success = pd.merge(stand_strategies, blackjack_success, left_on = "Strategy", right_on = "Strategy", how = "right")

# join start_values onto blackjack_success
blackjack_success = pd.merge(start_values, blackjack_success, left_on = "Hand", right_on = "Hand", how = "right")

# sort blackjack_success by index
blackjack_success = blackjack_success.sort_values(by = "index", ascending = True).reset_index(drop = True)

# remove the index column
blackjack_success = blackjack_success.drop("index", axis = "columns")

# compute the win percentage for each strategy of each player
strategy_score = []
for i in range(len(player_order) - 1):
    if i == 0:
        strategy_score.append(blackjack_success.groupby(player_order[i]).mean()[[player_order[i] + "_won"]].rename_axis("Player").reset_index())
    else:
        strategy_score.append(pd.concat([strategy_score[i - 1], blackjack_success.groupby(player_order[i]).mean()[[player_order[i] + "_won"]].reset_index(drop = True)], axis = "columns"))

# update strategy_score
strategy_score = strategy_score[len(player_order) - 2]

# compute the average strategy_score
strategy_score["Avg_won"] = strategy_score[[w + "_won" for w in player_order[:-1]]].apply(np.mean, axis = 1)

# sort strategy_score by Avg_won
strategy_score = strategy_score.sort_values(by = "Avg_won", ascending = False).reset_index(drop = True)

# compute the average table score
table_score = blackjack_success.groupby("Table").mean()["Table_won"].reset_index()

# sort table_score by Table_won
table_score = table_score.sort_values(by = "Table_won", ascending = False).reset_index(drop = True)

# compute the win percentage for each strategy of each player
strategy_hand_score = []
for i in range(len(player_order) - 1):
    if i == 0:
        strategy_hand_score.append(blackjack_success.groupby([player_order[i], player_order[i] + "_hand"]).mean()[[player_order[i] + "_won"]].rename_axis(["Player", "Player_hand"]).reset_index())
    else:
        strategy_hand_score.append(pd.concat([strategy_hand_score[i - 1], blackjack_success.groupby([player_order[i], player_order[i] + "_hand"]).mean()[[player_order[i] + "_won"]].reset_index(drop = True)], axis = "columns"))

# update strategy_hand_score
strategy_hand_score = strategy_hand_score[len(player_order) - 2]

# compute the average strategy_score
strategy_hand_score["Avg_won"] = strategy_hand_score[[w + "_won" for w in player_order[:-1]]].apply(np.mean, axis = 1)

# sort strategy_hand_score by Avg_won
strategy_hand_score = strategy_hand_score.sort_values(by = ["Avg_won", "Player"], ascending = False).reset_index(drop = True)

# get the best strategy for each starting hand value
hand_scores = pd.DataFrame(columns = strategy_hand_score.columns)
for i in list(set(strategy_hand_score["Player_hand"])):
    
    # get the scores for starting hand i
    hand_score_i = strategy_hand_score.loc[strategy_hand_score["Player_hand"] == i]
    
    # keep the max value(s) of Avg_won
    hand_score_i = hand_score_i.loc[hand_score_i["Avg_won"] == max(hand_score_i["Avg_won"])]
    
    # sort hand_score_i by Player and take the first row
    hand_score_i = hand_score_i.sort_values(by = "Player", ascending = True).reset_index(drop = True).head(1)
    
    # add hand_score_i to hand_scores
    hand_scores = pd.concat([hand_scores, hand_score_i], axis = "index").reset_index(drop = True)

# only keep hand values up to 21
hand_scores = hand_scores.loc[hand_scores["Player_hand"] <= 21]

# sort hand_scores by Player and Player_hand
hand_scores = hand_scores.sort_values(by = ["Player", "Player_hand"], ascending = True).reset_index(drop = True)

# write out hand_scores
hand_scores.to_csv("Blackjack Strategy.csv", index = False)











